book/compiler-backend/README.md

```diff
@@ -186,75 +185,86 @@ benchmarking]{.idx}
-    | Bob     -> 101 in
-  List.iter ~f:(fun v -> ignore(test v))
-    [ Alice; Bob ]
-
-let monomorphic_pattern_large () =
-  let test v =
-    match v with
-    | A       -> 100
-    | B       -> 101
-    | C       -> 102
-    | D       -> 103
-    | E       -> 104
-  in
-  List.iter ~f:(fun v -> ignore(test v))
-    [ A; B; C; D ]
-
-let tests = [
-  "Polymorphic pattern", polymorphic_pattern;
-  "Monomorphic larger pattern", monomorphic_pattern_large;
-  "Monomorphic small pattern", monomorphic_pattern_small;
-]
+module Monomorphic = struct
+  type t =
+    | Alice
+    | Bob
+    | Charlie
+    | David
+
+  let bench () =
+    let convert v =
+      match v with
+      | Alice -> 100
+      | Bob -> 101
+      | Charlie -> 102
+      | David -> 103
+    in
+    List.iter
+      ~f:(fun v -> ignore (convert v))
+      [ Alice; Bob; Charlie; David ]
+end
+
+module Monomorphic_small = struct
+  type t =
+    | Alice
+    | Bob
+
+  let bench () =
+    let convert v =
+      match v with
+      | Alice -> 100
+      | Bob -> 101
+    in
+    List.iter
+      ~f:(fun v -> ignore (convert v))
+      [ Alice; Bob; Alice; Bob ]
+end
+
+module Polymorphic = struct
+  type t =
+    [ `Alice
+    | `Bob
+    | `Charlie
+    | `David
+    ]
+
+  let bench () =
+    let convert v =
+      match v with
+      | `Alice -> 100
+      | `Bob -> 101
+      | `Charlie -> 102
+      | `David -> 103
+    in
+    List.iter
+      ~f:(fun v -> ignore (convert v))
+      [ `Alice; `Bob; `Alice; `Bob ]
+end
+
+let benchmarks =
+  [ "Monomorphic large pattern", Monomorphic.bench
+  ; "Monomorphic small pattern", Monomorphic_small.bench
+  ; "Polymorphic large pattern", Polymorphic.bench
+  ]
 
 let () =
-  List.map tests ~f:(fun (name,test) -> Bench.Test.create ~name test)
+  List.map benchmarks ~f:(fun (name, test) ->
+      Bench.Test.create ~name test)
   |> Bench.make_command
   |> Command.run
 ```
 
 Building and executing this example will run for around 30 seconds by
 default, and you'll see the results summarized in a neat table:
 
-```scheme file=examples/back-end/bench_patterns/dune
-(executable
-  (name      bench_patterns)
-  (modules   bench_patterns)
-  (libraries core_bench))
-```
-
-
-
 ```sh dir=examples/back-end/bench_patterns,non-deterministic=command
 $ dune exec -- ./bench_patterns.exe -ascii -quota 0.25
-Estimated testing time 750ms (3 benchmarks x 250ms). Change using '-quota'.
-
-  Name                         Time/Run   Percentage
- ---------------------------- ---------- ------------
-  Polymorphic pattern           15.72ns      100.00%
-  Monomorphic larger pattern     9.41ns       59.89%
-  Monomorphic small pattern      8.73ns       55.56%
-
+File "bench_patterns.ml", line 1:
+Error: The files /home/yminsky/Code/rwo/_build/install/default/lib/core/core.cmi
+       and /home/yminsky/Code/rwo/_build/install/default/lib/core_bench/core_bench__Bench.cmi
+       make inconsistent assumptions over interface Sexplib__Pre_sexp
```

Comment:
Something's gone wrong here

Reply:
Indeed. That's why this is a draft. runtest produced those promotions. I was hoping you might have some idea why!

Reply:
fixed in 38836a9e876ea5643e183d82e688a6b57c78fcb0 -- was missing .rwo-example files for those two so it was picking up your global ppx_sexp